lib/smart_proxy_dns_powerdns/backend/mysql.rb

```
@@ -35,15 +35,20 @@ def create_record domain_id, name, type, content
       name = connection.escape(name)
       content = connection.escape(content)
       type = connection.escape(type)
-      connection.query("INSERT INTO records (domain_id, name, ttl, content, type) VALUES (#{domain_id}, '#{name}', #{ttl.to_i}, '#{content}', '#{type}')")
+      connection.query("INSERT INTO records (domain_id, name, ttl, content, type, change_date) VALUES (#{domain_id}, '#{name}', #{ttl.to_i}, '#{content}', '#{type}', UNIX_TIMESTAMP())")
       connection.affected_rows == 1
     end
 
     def delete_record domain_id, name, type
       name = connection.escape(name)
       type = connection.escape(type)
       connection.query("DELETE FROM records WHERE domain_id=#{domain_id} AND name='#{name}' AND type='#{type}'")
-      connection.affected_rows >= 1
+      ret = connection.affected_rows >= 1
+      if ret
+        connection.query("UPDATE records SET change_date=UNIX_TIMESTAMP() WHERE domain_id=#{domain_id} AND type='SOA'")
+        ret = connection.affected_rows == 1
```

I think if this is not 1 then it should a log message. Some cases I can imagine:
* There is no SOA record: 0 affected rows --> problem)
* There are multiple SOA records: > 1 affected rows --> problem)
* The change date is already set to the current unix timestamp: 0 affected rows --> race condition, might be a warning but possible problem

In any of those cases it shouldn't affect the return value since a rectify-zone might still be needed in race condition case.

The same goes for the postgresql backend.

---

You are right, logging would definitely help! The thing is I am not a Rubyist and don't know much about the log interface of the SmartProxy. Any advice would be appreciated.

Also, I think the README needs to be updated to guide to the `SOA-EDIT` properties defined in PowerDNS. I honestly totally missed this part of the PDNS guide at first and this is a required setting in order to have a working, auto-incremented SOA serial.

Regarding your last usecase I don't this this is true: update will change the SOA record even if the current change_date is already set. My guess is that the backend will still return `affected_rows == 1` is this case because nothing in the SQL statement filters the SOA records with "up to date" change_date field.

---

Another question I have is: do we need to fail hard when update of the SOA fails?
My guess would be yes because usually, if the SOA serial does not get bumped, no other DNS slaves get notified. Also, forwarders would no catch up on the changes until the record TTL has expired (since no change in SOA serial means no change to the zone content).

Maybe raising a proper exception with plenty of information would be the best way to go.
All of this would need to be run inside a transaction though because this could lead to the following bad scenario otherwise:

* delete record → success
* update SOA → fails, raise an exception
* user tries again
* delete record →no record deleted → does not update SOA :(

---

> You are right, logging would definitely help! The thing is I am not a Rubyist and don't know much about the log interface of the SmartProxy. Any advice would be appreciated.

I'm not a Rubyist either so I'm doing the best I can, but the plugin logs very little. Currently there is only one example in this code base: https://github.com/theforeman/smart_proxy_dns_powerdns/blob/566cdec3c8d5335ee0cbecf6bdeb4c7a6488756d/lib/smart_proxy_dns_powerdns/backend/rest.rb#L87 and I'm not sure the extends is needed there since the main class also does that so first try without it.

> Also, I think the README needs to be updated to guide to the SOA-EDIT properties defined in PowerDNS. I honestly totally missed this part of the PDNS guide at first and this is a required setting in order to have a working, auto-incremented SOA serial.

Absolutely.

> Regarding your last usecase I don't think this is true: update will change the SOA record even if the current change_date is already set. My guess is that the backend will still return affected_rows == 1 is this case because nothing in the SQL statement filters the SOA records with "up to date" change_date field.

This is not true for MySQL. An ```UPDATE``` statement will set ```affected_rows``` to the number of rows that changed. If the column already contains the value then it will not increase the ```affected_rows```. Since unix timestamps have a resolution of 1 second there's a decent chance this will happen.

---

> Another question I have is: do we need to fail hard when update of the SOA fails?

That was my initial thinking as well, but the race condition means there's a valid use case where there is no update and we can't really detect. I'm not sure how to solve that.

---

> This is not true for MySQL. An UPDATE statement will set affected_rows to the number of rows that changed. If the column already contains the value then it will not increase the affected_rows. Since unix timestamps have a resolution of 1 second there's a decent chance this will happen.

This is really bad news :( We will have to discard the return value of this call then because there is a high chance of this happening when Foreman deletes multiple hosts of the same domain.
Also I am wondering if PowerDNS can serve a zone without SOA record anyway so the problem can be reduced to the "already up to date" race stated above.

Should I discard the return value of the UPDATE statement then?

---

> Should I discard the return value of the UPDATE statement then?
> Maybe, at least issuing a warning would be better than silently ignoring the result.

I think it should be an info message. The race condition is quite possible and I'd rather not have false positive warnings. The other cases aren't that likely and I suspect you'll notice the breakage in other ways. If there is no SOA then PowerDNS won't serve the zone and no idea what happens when you have multiple SOA records. Maybe that should be a warning.